master/internal/stream/system.go

```diff
@@ -255,72 +254,62 @@ func (ps *PublisherSet) Websocket(
 		}
 	}()
 
-	// reads is where we collect SubscriptionModMsg messages we read from the websocket until
-	// waitForSomething() delivers those messages to the websocket goroutine.
-	var reads []SubscriptionModMsg
-
-	// always be reading for new subscriptions
+	// monitor for new StartupMsgs
 	go func() {
-		defer streamer.Close()
 		for {
-			var mods SubscriptionModMsg
-			err := socket.ReadJSON(&mods)
+			select {
+			// check if we're canceled
+			case <-ctx.Done():
+				return
+			default:
+			}
+			var newStartup StartupMsg
+			err := socket.ReadJSON(&newStartup)
+			// XXX: errors here don't seem to appear on the websocket side...?
+			// check if error was due to something other than the connection closing.
+			if err != nil && !websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+				log.Errorf("error while reading additional startup message: %s", err)
+			}
 			if err != nil {
-				if websocket.IsUnexpectedCloseError(
-					err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure,
-				) {
-					log.Errorf("unexpected close error: %v", err)
-				}
-				break
+				streamer.Close()
+				return
+			}
+			resetMsgs, err := ps.performStreamerSoftReset(ctx, user, newStartup, streamer, &ss)
+			if err != nil {
+				log.Debugf("error during soft reset: %s\n", err)
+				return
+			}
```

Comment:
whew, this is a tricky place to put this.  Most of the streamer business logic is centralized in one goroutine, but here we're doing startup logic on another goroutine (but not all the time).

Is there a way to signal to the main goroutine, so the reset logic stays single-threaded?

Reply:
Yep! We could do something similar to what was being done with reads for mods

Reply:
You marked this as resolved, but you're still doing processing of the startup message on two different goroutines, depending on when it comes in.  That doesn't feel like my comment is resolved...?

Reply:
Just pushed up the commit!